File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if(nums.length==0){
            return 0;
        }
        int max=nums[0];
        int min=nums[0];
        int result=max;

        for(int i=1;i<nums.length;i++){
            int curr=nums[i];
            int tempmax=Math.max(curr,Math.max(curr*max,curr*min));
             min=Math.min(curr,Math.min(curr*max,curr*min));
             max=tempmax;

             result= Math.max(result,max);
        }
        return result;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
         if (nums.empty()) return 0;

        int maxProd = nums[0];
        int minProd = nums[0];
        int result = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            int curr = nums[i];
            int tempMax = max({curr, curr * maxProd, curr * minProd});
            minProd = min({curr, curr * maxProd, curr * minProd});
            maxProd = tempMax;

            result = max(result, maxProd);
        }
        return result;
    }
    
};